File: Earthshake/EarthshakeServiceFactory.m
//
//  EarthshakeServiceFactory.m
//  Earthshake
//
//  Created by Andrey Paunov on 2016-03-26.
//  Copyright © 2016 Andrey Paunov. All rights reserved.
//

#import "EarthshakeServiceFactory.h"
#import "EarthshakeModelFactory.h"
#import "EarthshakeServiceImpl.h"

@implementation EarthshakeServiceFactory
{
    id<EarthshakeService> earthshakeService;
}

- (id<EarthshakeService>) createEarthshakeService
{
    if (!earthshakeService)
    {
        earthshakeService = [[EarthshakeServiceImpl alloc] init]; // EarthshakeServiceImpl complies with the EarthshakeService protocol
        earthshakeService.earthshakeItemBuilder = [self.earthshakeModelFactory createEarthshakeItemBuilder];
    }

    return earthshakeService;
}

@end

File: Earthshake/EarthshakeServiceFactory.swift
//
//  EarthshakeServiceFactory.swift
//  Earthshake
//
//  Created by Andrey on 2016-12-08.
//  Copyright © 2016 AP. All rights reserved.
//

import Foundation

class EarthshakeServiceFactory : NSObject
{
    var earthshakeModelFactory: EarthshakeModelFactory?
    var earthshakeService: EarthshakeService?

    func createEarthshakeService() -> EarthshakeService?
    {
        if (earthshakeService == nil)
        {
            earthshakeService = EarthshakeServiceImpl()
            earthshakeService!.earthshakeItemBuilder = earthshakeModelFactory!.createEarthshakeItemBuilder()
        }

        return earthshakeService
    }
}
